### PR TITLE
Overseer: subsystems communicate directly

### PR DESCRIPTION
**Pull request by rphmeier**
The previous behavior was that messages would go first to the overseer and then be relayed onwards to the destination subsystem.

Now, subsystems communicate via channels directly to other subsystems. Additionally, they have the option of either sending a message through a bounded or unbounded channel. The unbounded channels should be used sparingly in situations where:
  1. There is a communication cycle between subsystems
  2. One of the parts of the cycle has a clear bound on the number of messages produced.

A few examples of places we might use `send_unbounded_message` are when we send locally generated backing statements to statement distribution or when we send assignments and approvals from approval voting to the approval distribution subsystem. This avoids deadlock when a network subsystem is `.await`ing a send to the core subsystem, while the core subsystem sends to the networking subsystem at the same time. If buffers are full, that will deadlock. Unbounded channels avoids this situation.

`SubsystemContext` and the overseer itself have an invariant that when a subsystem sends a message after receiving a signal from the overseer, that the recipient of the message will not receive it until having observed the same signal. This allows subsystems to assume that other subsystems are aware of the same view of blocks as they are.

My change-set here upholds the same invariant by having each subsystem tag each message with the number of signals that it had observed before sending the message. If the recipient has not received at least that many signals, it pockets the message until it has received enough signals. This rests on the fact that the overseer sends signals to all subsystems equally. If one subsystem receives a signal, others can expect to receive it shortly.

I have also given `SubsystemContext` a `SubsystemSender` assoc. type which will be used to split sending and receiving. Follow-up PRs will make use of this feature.

### CODE REVIEW

File: node/overseer/src/lib.rs
Comment (coriolinus) [inline node/overseer/src/lib.rs:346]: It's too bad we don't have HKTs yet; if we did, you could simplify this into

```rust
let make_packet = for<T> |message: T| { ... }
```

and simply capture the values of `timer` and `signals_received`, which would simplify usage.

File: node/overseer/src/lib.rs
Comment (coriolinus) [inline node/overseer/src/lib.rs:404]: I suspect that at some point we're going to want to debug which particular subsystem was the intended recipient of the failed message. Maybe an api like 

```rust
impl AllMessages {
  pub fn recipient_name(&self) -> &'static str { ... }
}
```

would make that simple to add to this debug message.

**Comment by eskimor** [comment]
> Our use of channels is still likely to lead to deadlocks, because we `.await` on both sides unconditionally. Future refactoring should alter subsystems to time-out messages to other subsystems so that progress can be made even under heavy load.

I think we could do better than this, but I need to check a few things first and think a bit more about it.

**Review by coriolinus — APPROVED** [review]
Core logic in `OverseerSubsystemContext::recv` LGTM. I agree that this maintains the message invariant.

I'm pleasantly surprised that making this change didn't require changing any of the individual subsystems.

**Review by eskimor — APPROVED** [review]
Looks sensible to me.

File: node/overseer/src/lib.rs
Comment (eskimor) [reply node/overseer/src/lib.rs:346]: Well we can do without, if we are sure this is the way to go, we can just expose the subsystem senders directly.

**Review by drahnr — APPROVED** [review]
Implementation wise, it looks good.

As a side effect here we lose the overseer as our probe point for messages though.

File: node/overseer/src/lib.rs
Comment (rphmeier) [reply node/overseer/src/lib.rs:346]: I had the same thought when I was writing the code, but unfortunately no HKT yet

File: node/overseer/src/lib.rs
Comment (ordian) [inline node/overseer/src/lib.rs:1363]: the docs here are outdated

File: node/subsystem/src/lib.rs
Comment (ordian) [inline node/subsystem/src/lib.rs:231]: feels like it's missing a motivation for this change and a guide how it's supposed to be used, basically turns comments in Elements into the docs

Comment (rphmeier) [reply node/subsystem/src/lib.rs:231]: We have discussed motivation, no? There's no way to solve the deadlock without having tools for breaking cycles.

Comment (ordian) [reply node/subsystem/src/lib.rs:231]: I meant putting the motivation in the doc/mod comments for future maintainers/new joiners. It's not possible to hold every Element conversation in head over a long period of time.

File: node/overseer/src/lib.rs
Comment (ordian) [inline node/overseer/src/lib.rs:1773]: naming nit: what's the difference between `parachain_subsystem` and `parachain_overseer` prefix? They both are parameterized by `subsystem_name`.

File: node/overseer/src/lib.rs
Comment (ordian) [inline node/overseer/src/lib.rs:1467]: tracking of the difference between received signal across subsystems adds a lot of complexity and not sure if it actually solves any problems

if a subsystem is busy processing messages and doesn't catch up with signals, it might be awaiting a message from another subsystem and sending it a bunch of signals wouldn't unblock it 

what we can do alternatively is use select_biased to prefer selecting signals to messages

Comment (rphmeier) [reply node/overseer/src/lib.rs:1467]: > tracking of the difference between received signal across subsystems adds a lot of complexity and not sure if it actually solves any problems

We don't do this, but yeah, I don't think that solves any problems.

I can switch to select_biased. But I believe that's not necessary as select will choose randomly if both streams are ready. The probability of selecting only messages is vanishingly small.

We use send_unbounded to break cycles so that awaiting on a subsystem won't stall. Although we will encode this in the type system in a future refactoring

Comment (ordian) [reply node/overseer/src/lib.rs:1467]: > We don't do this

Maybe I named it poorly, but that's what I meant we do with `pending_incoming`.

Comment (rphmeier) [reply node/overseer/src/lib.rs:1467]: We don't track any difference. What exactly is the issue you have with `pending_incoming`? I added that code to maintain the invariant that a message sent after a signal is received on subsystem A will only arrive after the same signal on subsystem B. This is an important property which our subsystems rely on heavily, so I won't remove it.

File: node/overseer/src/lib.rs
Comment (rphmeier) [reply node/overseer/src/lib.rs:1773]: `subsystem_signals_sent` sounds like the subsystems are sending the signals but they are actually receiving them. `overseer_unbounded_sent` has the same issue in the opposite direction.

Comment (rphmeier) [reply node/overseer/src/lib.rs:1773]: These are `OverseerSignals` so `overseer_signals_sent` makes sense I think. But I could rename `subsystem_bounded_sent` etc.. not sure to what

File: node/subsystem/src/lib.rs
Comment (rphmeier) [reply node/subsystem/src/lib.rs:231]: ok, added

File: node/overseer/src/lib.rs
Comment (ordian) [reply node/overseer/src/lib.rs]: > What exactly is the issue you have with pending_incoming?

Added complexity. I believe it can be achieved in a more simple way, e.g. with a 
https://docs.rs/futures/0.3.13/futures/lock/struct.Mutex.html
locked in `broadcast_signal` and `route_message`, can even be an `RwLock` with `write` in `broadcast_signal` and `read` in `route_message`. This will guarantee the ordering between signals and messages now that we use `select_biased!`.

One might argue that it will block overseer if the signal channel buffer is full for some subsystem, but in this impl it will indefinitely block the subsystem.

Comment (ordian) [reply node/overseer/src/lib.rs]: although that doesn't work after we add direct communication, yeah, ok

File: node/subsystem/src/lib.rs
Comment (ordian) [inline node/subsystem/src/lib.rs:292]:
```suggestion
	/// Generally, for this method to be used, these conditions should be met:
	/// * There is a communication cycle between subsystems
	/// * One of the parts of the cycle has a clear bound on the number of messages produced.
```

**Comment by rphmeier** [comment]
bot merge

**Review by eskimor — APPROVED** [review]

**Pull request merged**
